File: Utilities/webDeploy.py
#!/bin/python
import netifaces
import random
import sys

def genRandomFileName():
    usableCharacters = "012345679ABCDEF"

    randomFileName = ""
    for i in range(32):
        randomFileName += random.choice(usableCharacters)

    return(randomFileName)

def printUrls(webPath):
    addresses = []
    for iface in netifaces.interfaces():
        if(iface != "lo"):
            addresses.append(netifaces.ifaddresses(iface))

    for addr in addresses[0]:
        ip = addresses[0][addr][0]["addr"]
        print("http://" + ip + "/" + webPath)

def readFile(inFilePath):
    try:
        with open(inFilePath, "rb") as inFile:
            return inFile.read()
        
    except FileNotFoundError:
        print("Error reading " + inFilePath + ": File not found. Please try again.")
        exit
    except PermissionError:
        print("Error reading " + inFilePath + ": Insufficient permissions. Please try again.")
        exit
    except:
        print("Error reading " + inFilePath + ": Unkown error. Please try again.")
        exit

def writeFile(inFilePath, outFileName):
    outFilePath = "/var/www/html/" + outFileName
    
    try:
        with open(outFilePath, "wb") as outFile:
            outFile.write(readFile(inFilePath))
    
    except PermissionError:
        print("Error writing " + outFilePath + ": Insufficient permissions. Please try again.")
        exit
    except:
        print("Error writing " + outFilePath + ": Unkown error. Please try again.")
        exit



if(len(sys.argv) == 1):
    print("\nError: No arguments provided. Please provide file to be posted as argument.")
    print("Example: ./" + sys.argv[0] +" [Input file path]\n")
    exit

else:
    inFilePath = sys.argv[1][0]
    outFileName = genRandomFileName()
    writeFile(inFilePath, outFileName)

    print("\n\nYour file is hosted at these links:")
    printUrls(outFileName)
    print("\n\n")

File: Utilities/webDeploy.cpp
#include <ctime>
#include <fstream>
#include <iostream>

//Moves a file to "/var/www/html/" to be hosted by Apache for remote deployment
//intended for use on linux systems 
//Networking and Apache configuration are not performed by this program

//Generates a random hex string to use as the file name
std::string genRandomName()
{
    std::srand(std::time({}));

    char usableCharacters[] = "0123456789ABCDEF";

    std::string randomName;

    int nameLength;
    nameLength = 32;
    for(int i = 0; i < nameLength; i++)
    {
        randomName += usableCharacters[std::rand() % 16];
    }

    std::string outFilePath;
    outFilePath = "/var/www/html/" + randomName;
    std::cout << "\nyour file is now hosted at:\n";
    std::cout << "\nhttp://[your-IP-Address]/" << randomName << "\n\n"; 
    
    return outFilePath;
}

std::ifstream readFile(std::string fileName)
{
    std::ifstream inFile{fileName};

    if(!inFile)
    {
        std::cerr << "Could not open " << fileName << " for reading.\n";
        std::cerr << "Please try again.\n";
        std::cerr << "Exiting.\n";
        exit;
    }
    
    return inFile;
}

int writeFile(std::string inFileName, std::string outFilePath)
{

    std::ifstream inFile{readFile(inFileName)};
    
    std::ofstream outFile{outFilePath};

    if(!outFile)
    {
        std::cerr << "Could not open " << outFilePath << " for writing.\n";
        std::cerr << "Please try again with higher permissions.\n";
        return 1;
    }

    std::string currentLine;
    while(std::getline(inFile, currentLine))
    {
        outFile << currentLine << "\n";
    }

    return 0;
}



int main()
{

    std::string fileName;
    std::cout << "Enter file to post:\n";
    std::cin >> fileName;

    std::string outPath;
    outPath = genRandomName();

    if(writeFile(fileName, outPath) == 0)
    {
        return 0;
    }
    else
    {
        return 1;
    }
}
